### ISSUE TITLE
Moving protobuf definitions out of the ocis repo

### ISSUE THREAD

**Issue by jvillafanez** [owncloud/ocis#2822]
## Is your feature request related to a problem? Please describe.
Protobuf definitions are intended to be implemented in multiple languages or by multiple unrelated application. Having the definitions inside the ocis repo doesn't respect this idea since they'll be more difficult to access and to implement in other languages.

## Describe the solution you'd like
The protobuf definitions will be stored in a separated repository ("owncloud/ocisproto" for example). The ocis repo will only contain the minimum information to generate the code for the services.
Note that each ocis microservice should generate its own generated code, or place the generated code in a common location, such as the "ocis-pkg" folder. Using the generated code in a microservice by a different microservice shouldn't happen.

### Folder structure for the protobuf repository

```
.
├── buf.lock
├── buf.yaml
├── msgtypes
│   └── v1
│       ├── accounts.proto
│       └── config.proto
├── ocisservices
│   └── v1
│       ├── accounts.proto
│       └── config.proto
└── README.md
```
The msgtypes folder will contain only "Message" definitions. These are intended to be used as data types that could be reused in different places.
For example, in the msgtypes/v1/accounts.proto, you can define a message "Account" that could be reused not just in the accounts service but in any other service that could need it.

The ocisservices folder will contain the "Service" definitions. Each service should define a "<methodName>Request" and a "<methodName>Response" messages for each rpc method in each service. These messages are expected to be defined in the proto file along with the service definition. This is mainly to conform with the Buf lint requirements.

This structure allows API versioning in case we need to perform breaking changes in the API. It's also easy to distinguish what the users of these protobuf definitions should implement (one or more ocisservices)

#### buf.yaml file

No changes comparing with the current file.
```
version: v1
lint:
  use:
    - DEFAULT
deps:
  - buf.build/googleapis/googleapis
breaking:
  use:
    - FILE
```

### Code generation

Inside the ocis repo, we'll place a "buf.gen.yaml" file with content similar to the following:
Note that this is for the accounts microservice
```
version: v1
managed:
  enabled: true
  go_package_prefix:
    default: github.com/owncloud/ocis/accounts/pkg
    except:
      - buf.build/googleapis/googleapis
plugins:
  - name: go
    path: ../.bingo/protoc-gen-go
    out: pkg/
    opt:
      - paths=source_relative

  - name: micro
    path: ../.bingo/protoc-gen-micro
    out: pkg/
    opt:
      - paths=source_relative

  - name: microweb
    path: ../.bingo/protoc-gen-microweb
    out: pkg/
    opt:
      - paths=source_relative

  - name: openapiv2
    path: ../.bingo/protoc-gen-openapiv2
    out: pkg/

  - name: doc
    path: ../.bingo/protoc-gen-doc
    out: ../docs/extensions/accounts
    opt:
      - ./templates/GRPC.tmpl,grpc.md
```
This is the only piece we need inside the ocis repo. The actual content and the placement will depend on how we want to handle the generated code.

The initial idea was to place the "buf.gen.yaml" file in each microservice that needs to generate a client or server for the service. However, this might cause a lot of duplication in the ocis repo because the code generated for the same service is expected to be the same.

A better idea could be to place it inside the "ocis-pkg" and generate all the code there. This way, all the microservices could use the code generated there. (We'll assume that each microservice will generate its own code for the following examples. Changes will be needed if we want to generate the code in the "ocis-pkg" folder)

The code will be generated with the following command:
```
buf generate https://github.com/<user>/<repo>.git --path ocisservices/v1/accounts.proto
```

Note that the code will be generated in "pkg/ocisservices/v1", not in "pkg/proto/v0".

In case we need more services, they can be generated the same way:
```
buf generate https://github.com/<user>/<repo>.git --path ocisservices/v1/accounts.proto
buf generate https://github.com/<user>/<repo>.git --path ocisservices/v1/config.proto
```

To generate all the services (v2, v3 and others will also be generated):
```
buf generate https://github.com/<user>/<repo>.git --path ocisservices/
```

Also note that the generated files will be all inside the same folder, as shown below. This could get messy if the microservice requires too much communication.
```
pkg/ocisservices/
└── v1
    ├── accounts.pb.go
    ├── accounts.pb.micro.go
    ├── accounts.pb.web.go
    ├── accounts.swagger.json
    ├── config.pb.go
    ├── config.pb.micro.go
    ├── config.pb.web.go
    └── config.swagger.json
```

### Alternative folder structure
```
.
├── buf.lock
├── buf.yaml
├── msgtypes
│   ├── accounts
│   │   └── v1
│   │       └── types.proto
│   └── config
│       └── v1
│           └── types.proto
├── ocisservices
│   ├── accounts
│   │   └── v1
│   │       └── accounts.proto
│   └── config
│       └── v1
│           └── config.proto
└── README.md
```

This will generate a different folder structure:
```
pkg/ocisservices/
├── accounts
│   └── v1
│       ├── accounts.pb.go
│       ├── accounts.pb.micro.go
│       ├── accounts.pb.web.go
│       └── accounts.swagger.json
└── config
    └── v1
        ├── config.pb.go
        ├── config.pb.micro.go
        ├── config.pb.web.go
        └── config.swagger.json
```

**Comment by refs** [comment]
Thanks for the write up! I have a few concerns. In your example you're only referring to these 2 plugins:

```yaml
plugins:
  - name: go
    path: ../.bingo/protoc-gen-go
    out: pkg/
    opt:
      - paths=source_relative

  - name: micro
    path: ../.bingo/protoc-gen-micro
    out: pkg/
    opt:
      - paths=source_relative
```

but how would that scale with multiple languages? To quote your second folder structure (which I favor), it *could* look similar to:

```
pkg/ocisservices/
├── accounts
│   └── v1
│     └──  go
│          ├── accounts.pb.go
│          ├── accounts.pb.micro.go
│          ├── accounts.pb.web.go
│          └── accounts.swagger.json
```

And the single repository would be responsible to holding the definitions in every language. But I would like to take a step back and discuss the first paragraph:

> Protobuf definitions are intended to be implemented in multiple languages or by multiple unrelated application. Having the definitions inside the ocis repo doesn't respect this idea since they'll be more difficult to access and to implement in other languages.

It is the second sentence the one I'm having troubles wrapping my head around. The case here being tooling, tools like `buf` just need a `.proto` file, and it just so happen that our files are ... public, so if you're in the need of a client in C#, you can just use the protobuf definitions with the proto file on our public repo. While I agree is far from ideal, it is not necessarily "more difficult to access". Now let's look at the other side, say you want to use the `accounts.proto` file to generate an accounts server in C#, for which case the accounts.proto probably has some dependencies on another proto file, but the procedure is the same, you refer to the proto file, generate the **interfaces** and implement them. [There is a good read about the matter](https://medium.com/stackpulse/grpc-in-practice-directory-structure-linting-and-more-d4d438ac4f86) and why they chose not to put all the proto files into a single repository, and do it the way we do it, citing the benefits being:

1. Modifying a proto file is just like modifying the code. It’s just a file inside the repository and it can be done rapidly during microservice development.
2. Proto files and their compilation results are versioned together with the code that implements the services.
 3. If service1 wants to talk to service2 using a gRPC client, it will use go modules to fetch service2‘s compiled proto messages and clients. It can reference the auto-generated client code and the message structs to easily communicate with service2 using native Go tools.

**Comment by jvillafanez** [comment]
From my understanding, there are 3 components: the service definition, the generated client code and the generated server code. We can join server and client generated code into one because that's what the code generator does (I don't like it, but we can't change it)

Let's assume I want to create an app to create avatars, and I need to access to the thumbnail microservice. Note that my app is in my own repo, which is in PHP (or any other language). How could we do this?
It doesn't look good to clone the ocis repo to get the protobuf definition because it contains a lot of other things, and we cannot force ocis (or anyone) to generate PHP code.

Going back to the components topic, I think this proposal makes more sense in the way it's more clear who's responsible of what.
The "ocisproto" repo will contain only the service definitions so anyone can easily get them and generate the code needed.
ocis will act as any other 3rd party microservice: get the definitions from the repo and generate the code it needs, just what it needs.
My app would do the same as ocis, with the difference that my app would generate PHP code. In any case, it's my responsibility to generate the PHP code, not ocis'

> but how would that scale with multiple languages? To quote your second folder structure (which I favor), it could look similar to:

In general, each microservice should take care of that. If my microservice is implemented in go, I want to generate go code. If it's implemented in PHP, I'll generate PHP code.
All ocis' microservices are implemented in go, so there is no need for ocis to generate code for more languages.

For the general use case, you only need to generate code for the language you're using in your app. You might want to generate additional stuff as seen in ocis, but it should be very limited. I don't think we need to scale it.

For the case of ocis, I think it's ok that ocis as a whole generates the go code for all the microservices that are in the ocis repo, and store the generated code in the "ocis-pkg" as a common ground.

> And the single repository would be responsible to holding the definitions in every language.

The definitions are language-agnostic. In fact, the definitions in the repo aren't expected to contain any language-specific option because you don't know what languages that definition will be used for.

The managed mode of Buf can take care of adding the language-specific options during the code generation, and the code generation will happen on the app's (ocis, in this particular case) repo


>    Modifying a proto file is just like modifying the code. It’s just a file inside the repository and it can be done rapidly during microservice development.
>    Proto files and their compilation results are versioned together with the code that implements the services.
>    If service1 wants to talk to service2 using a gRPC client, it will use go modules to fetch service2‘s compiled proto messages and clients. It can reference the auto-generated client code and the message structs to easily communicate with service2 using native Go tools.

I think this depends on whether we want to allow the creation of 3rd party microservices or not. If I'm allowed to create my own microservice, I don't think I'll get benefit of any of those points.
Unless we discard 3rd party microservices, I'd prefer to follow the same path as any outsider would go. This way we can say "This is how we're doing this and you're expected to do the same"

If we don't care about 3rd party microservices and we'll handle everything, it's ok to skip the external repo for the protobuf definitions and keep them inside the ocis repo. In any case, I think it's good to keep the distinction between the protobuf definition and the generated code.

**Comment by refs** [comment]
But I don't see any issues with defining the packages in the proto file. In fact, [that is the encouraged way of working with proto files](https://developers.google.com/protocol-buffers/docs/reference/go-generated#package)

> We recommend declaring it within the .proto file so that the Go packages for .proto files can be centrally identified with the .proto files themselves and to simplify the set of flags passed when invoking protoc. If the Go import path for a given .proto file is provided by both the .proto file itself and on the command line, then the latter takes precedence over the former. 

It does not clutter anything, it delegates less to the tooling (buf, or go protoc plugin) and makes things more explicit.

**Comment by jvillafanez** [comment]
It depends on who will take care of what responsibility. There are multiple points views on how to plan this topic depending on the goal.

If the goal is to make a library in multiple languages to connect to our services, it make sense your proposal. The problem is that I think it's too much work to do this properly because you'd need to take care of languages such as C++, PHP, Java and others. It would be unfair to say "we provide this particular client for Go, but you're on your own with PHP, Java, etc"
You'll also likely need to include some kind of testing for all the generated code.

In this case, you'll own and take care of everything, so it make sense to put language-specific options in the proto files because you expect nobody else to use them. Any 3rd party app (as well as ocis) is expected to use the generated code in any of the languages we're generating the code from. The service definition itself is internal and closed (in terms of expected usage, the code itself is expected to be open). If anyone could implement and generate code from our proto files, I don't think we should take that burden.

My point is, since we don't want to take the burden of taking care of all the languages supported, we'll just make the proto files public so anyone can generate the code from those files. In this case, ocis is just another 3rd party app using those proto files to generate the code it needs.
Now, if we move the responsibility of generating the code to the 3rd party app, why would we try to force package names in a code we don't own? If this were the case, it seems better to create a library and let the 3rd party app to include and use it.

I'd prefer to see ocis as any other app so we can be as homogeneous as possible in this regard. I don't think it's a good idea to treat ocis as special while other apps could struggle or be forced to implement things in a different way.

Note that I think this approach is also homogeneous for any possible private / paid extension for ocis. The extension would just follow the same approach of ocis.

**Comment by jvillafanez** [comment]
After playing around a bit with the code generation, I've detected some problems, so we might need some additional restrictions.

If we aren't going to use the BSR that "Buf" provides, I think the simplest option is the following:

The directory structure will be the one below:
```
.
├── buf.lock
├── buf.yaml
├── ocis
│   ├── accounts
│   │   └── v1
│   │       └── accounts.proto
│   └── config
│       └── v1
│           └── config.proto
└── README.md
```
This means:
* one self-contained file per service.
* no dependencies

Splitting into multiple files will generate several files. For example, if we split the "accounts.proto" into "accounts.proto" and "types.proto" (both files inside the "ocis/accounts/v1" folder), the result would be:
```
accounts.pb.go        accounts.swagger.json  types.pb.web.go
accounts.pb.micro.go  types.pb.go            types.swagger.json
accounts.pb.web.go    types.pb.micro.go
```
We'd also need to adjust the "\*.web.go" files because there might be some unused imports. (it seems to be a bug in the protoc-gen-web tool)
It seems questionable to have the "types.pb.micro.go" and "types.pb.web.go" around, but they might be needed. Note that the without the "types*" files the code wouldn't compile due to missing definitions of types in the "accounts.*" files.

Things might be worse with external proto files such as what was proposed initially because we need to generate the code for those external files as well, and it's likely difficult to track what files we need to get.

This is why we'll simplify the folder structure: each service will have only one self-contained proto file.

With this approach, the code will be generated with: (for that specific service)
```
buf generate https://github.com/<user>/<repo>.git --path ocis/config/v1
```

the "buf.gen.yaml" (in the ocis-pkg folder) will be:
```
version: v1
managed:
  enabled: true
  go_package_prefix:
    default: github.com/owncloud/ocis/ocis-pkg
    except:
      - buf.build/googleapis/googleapis
plugins:
  - name: go
    path: ../.bingo/protoc-gen-go
    out: ./
    opt:
      - paths=source_relative

  - name: micro
    path: ../.bingo/protoc-gen-micro
    out: ./
    opt:
      - paths=source_relative

  - name: microweb
    path: ../.bingo/protoc-gen-microweb
    out: ./
    opt:
      - paths=source_relative

  - name: openapiv2
    path: ../.bingo/protoc-gen-openapiv2
    out: ./

  - name: doc
    path: ../.bingo/protoc-gen-doc
    out: ../docs/extensions
    opt:
      - ../accounts/templates/GRPC.tmpl,grpc.md,source_relative  <--- template should be out of the accounts folder
```
Note that we'll have to adjust the docs folder structure.

-----

Using the BSR might give us better options since the dependency management should be easier, but we need to setup an account and we'll likely need to setup CI to push to changes to the BSR (we need to use the same account).

Better check first if we're willing to use it and to setup additional stuff before investigating further.

**Comment by refs** [comment]
❤️ I see the benefits of using BSR, on the other hand I'm worried there is some vendor locking there. I think we have no complicated dependencies we have to import and don't think we need BSR at all. There are some shared google protobuf definitions somewhere, but we can always just add the `third_party` folder.

**Comment by jvillafanez** [comment]
I think I'll change the approach to plan it like an internal library. The main problem I've seen with the previous approach is that you'll have to generate code for all the dependent proto files on your own, the code generator doesn't follow these dependencies.

For example, if file A (service) imports B (data types) you have to generate code for both A and B on your own. This will become a problem when the project grows and it isn't clear which files we need in order to generate code for a specific service. Note that generating code only for A will make the compilation fail due to missing code, or it might use a different file, which could be worse.

By planning like an internal library, this isn't a problem because we'll always generate code for all the files. Ocis and any 3rd party app will import the generated go code as with any library.

Some additional notes:
* I'm checking https://github.com/owncloud/protoc-gen-microweb/issues/8 . We'll need to fix that issue in order to be able to split the proto files
* The generated code from the proto file will stay in the same repo that the proto files
  * To be decided the actual repo. A different repo might still have sense.
  * We'll only generate code for go. I don't think we can take care for other languages.
* 3rd party apps in different languages will have to use HTTP. We won't generate code for other languages, and I think it will be difficult for 3rd party apps to generate code on their own.

**Comment by jvillafanez** [comment]
Final approach:
```
.
├── docs
│   ├── CONFIGURATION.tmpl
│   ├── extensions (content auto-generated)
│   │   └── ocis
│   │       ├── messages
│   │       │   ├── accounts
│   │       │   │   └── v1
│   │       │   │       └── grpc.md
│   │       │   └── config
│   │       │       └── v1
│   │       │           └── grpc.md
│   │       └── services
│   │           ├── accounts
│   │           │   └── v1
│   │           │       └── grpc.md
│   │           └── config
│   │               └── v1
│   │                   └── grpc.md
│   └── GRPC.tmpl
├── gen
│   ├── go.mod
│   ├── go.sum
│   └── ocis (content auto-generated)
│       ├── messages
│       │   ├── accounts
│       │   │   └── v1
│       │   │       ├── accounts.pb.go
│       │   │       ├── accounts.pb.micro.go
│       │   │       ├── accounts.pb.web.go
│       │   │       └── accounts.swagger.json
│       │   └── config
│       │       └── v1
│       │           ├── config.pb.go
│       │           ├── config.pb.micro.go
│       │           ├── config.pb.web.go
│       │           └── config.swagger.json
│       └── services
│           ├── accounts
│           │   └── v1
│           │       ├── accounts.pb.go
│           │       ├── accounts.pb.micro.go
│           │       ├── accounts.pb.web.go
│           │       └── accounts.swagger.json
│           └── config
│               └── v1
│                   ├── config.pb.go
│                   ├── config.pb.micro.go
│                   ├── config.pb.web.go
│                   └── config.swagger.json
├── proto
│   ├── buf.gen.yaml
│   ├── buf.lock
│   ├── buf.yaml
│   └── ocis
│       ├── messages
│       │   ├── accounts
│       │   │   └── v1
│       │   │       └── accounts.proto
│       │   └── config
│       │       └── v1
│       │           └── config.proto
│       └── services
│           ├── accounts
│           │   └── v1
│           │       └── accounts.proto
│           └── config
│               └── v1
│                   └── config.proto
└── README.md
```
* New repository containing the proto files and the generated code from those proto files
  * Auto-generated content via "Buf" is marked in the folder structure above. There could be additional files needed in order to configure stuff
  * Proto files will contain the "go_package" option. There are no plans to include more languages.
* Planned changes in the proto files (to be checked how far we want to go on each item):
  * Split the proto files as shown in the directory structure
  * Follow "Buf lint" practices. This will likely break the current api
    * "v0" references should be replaced by "v1". This will need additional changes in some parts of ocis.
    * Some return values from the services should be wrapped in a "methodRequest" / "methodResponse" message instead of returning an object.

To be decided:
* Pinned tools to be used.
  * We'll need to move some tools from the ocis repo to the new one. Mainly "buf" and all the protoc-gen-* stuff we're using.
  * To be checked if we want to update any of them (other than the updated protoc-gen-microweb, I don't think we need to update anything)

Expected changes in the ocis repo:
* Most of the changes will be around adjusting the import path for the generated code in order to use the external library.
* Taking into account that the content of the proto files will be split, we might need to review some objects since they might not be found using the new generated code (for example, the "Account" object won't be under the accounts services but in the accounts messages). It should be easy to spot since the code should fail to compile otherwise.
* Some web apis are likely to change. For example, the `api/v0/accounts/accounts-list` might need to be adjusted to `api/v1/accounts/accounts-list` for consistency with the "buf lint" rules. Some js files (and maybe others) might need to change the endpoint

**Comment by jvillafanez** [comment]
Quick status update:
* Changes are being implemented in the proto_refactor branch https://github.com/owncloud/ocis/tree/proto_refactor
* The accounts and thumbnails proto files have been migrated to the protogen folder in the ocis repo:
  * generated protobuf files are committed in the branch
  * grpc.md generated are also committed
* Last commit in the branch (9a533c9029831eed351e53f33d9c61b7d512621d) should work, but you might need https://github.com/owncloud/protoc-gen-microweb/pull/10 in order to generate again the files for the thumbnails
* Import paths for both accounts and thumbnails services have been adjusted to use the common location in the protogen folder.
* There hasn't been any big changes in the proto files other than splitting the files.

To do things:
* Adjust bingo files to use https://github.com/owncloud/protoc-gen-microweb/pull/10 (the files were generated with a build with that PR)
* Adjust Makefiles. Currently, the files are generated manually using `buf generate` inside the "protogen/proto" folder (remember that https://github.com/owncloud/protoc-gen-microweb/pull/10 is required to generate the thumbnails service)
* There are other services to be migrated.

**Comment by wkloucek** [comment]
@jvillafanez this should be closed by #2960?
